app/consumer/genesis.go

```diff
@@ -74,19 +99,205 @@ func transform(jsonRaw []byte, ctx client.Context) (json.RawMessage, error) {
+	srcConGen := consumerTypes.GenesisState{}
+	err := ctx.Codec.UnmarshalJSON(jsonRaw, &srcConGen)
+	if err != nil {
+		return nil, fmt.Errorf("reading consumer genesis data failed: %s", err)
+	}
+
+	// remove retry_delay_period from 'params' if present (introduced in v4.x)
+	params, err := ctx.Codec.MarshalJSON(&srcConGen.Params)
+	if err != nil {
+		return nil, err
+	}
+	paramsMap := map[string]json.RawMessage{}
+	if err := json.Unmarshal(params, &paramsMap); err != nil {
+		return nil, fmt.Errorf("unmarshalling 'params' failed: %v", err)
+	}
+	_, exists := paramsMap["retry_delay_period"]
+	if exists {
+		delete(paramsMap, "retry_delay_period")
+	}
+	params, err = json.Marshal(paramsMap)
+	if err != nil {
+		return nil, err
+	}
+
+	// marshal GenesisState and patch 'params' value
+	result, err := ctx.Codec.MarshalJSON(&srcConGen)
+	if err != nil {
+		return nil, err
+	}
+	genState := map[string]json.RawMessage{}
+	if err := json.Unmarshal(result, &genState); err != nil {
+		return nil, fmt.Errorf("unmarshalling 'GenesisState' failed: %v", err)
+	}
+	genState["params"] = params
+
+	provider, err := ctx.Codec.MarshalJSON(&srcConGen.Provider)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshalling 'Provider' failed: %v", err)
+	}
+	providerMap := map[string]json.RawMessage{}
+	if err := json.Unmarshal(provider, &providerMap); err != nil {
+		return nil, fmt.Errorf("unmarshalling 'provider' failed: %v", err)
+	}
+
+	// patch .initial_val_set form .provider.initial_val_set if needed
+	if len(srcConGen.Provider.InitialValSet) > 0 {
+		valSet, exists := providerMap["initial_val_set"]
+		if !exists {
+			return nil, fmt.Errorf("'initial_val_set' not found in provider data")
+		}
+		_, exists = genState["initial_val_set"]
+		if exists {
+			genState["initial_val_set"] = valSet
+		}
+	}
+
+	// patch .provider_consensus_state from provider.consensus_state if needed
+	if srcConGen.Provider.ConsensusState != nil {
+		valSet, exists := providerMap["consensus_state"]
+		if !exists {
+			return nil, fmt.Errorf("'consensus_state' not found in provider data")
+		}
+		_, exists = genState["provider_consensus_state"]
+		if exists {
+			genState["provider_consensus_state"] = valSet
+		}
+	}
+
+	// patch .provider_client_state from provider.client_state if needed
+	if srcConGen.Provider.ConsensusState != nil {
+		clientState, exists := providerMap["client_state"]
+		if !exists {
+			return nil, fmt.Errorf("'client_state' not found in provider data")
+		}
+		_, exists = genState["provider_client_state"]
+		if exists {
+			genState["provider_client_state"] = clientState
+		}
+	}
+
+	// delete .provider entry (introduced in v3.3.x)
+	delete(genState, "provider")
+
+	// Marshall final result
+	result, err = json.Marshal(genState)
+	if err != nil {
+		return nil, fmt.Errorf("marshalling transformation result failed: %v", err)
+	}
+
+	// remove all prehash_key_before_comparison entries not supported by ics23 versions used by v2.x
```

Comment:
going the shortcut here bypassing json result manipulation as element is deeply nested in consumer genesis type.